infra/terraform/global/assets/create_etcd_ebs_snapshot/create_etcd_ebs_snapshot.go

```diff
@@ -0,0 +1,150 @@
+package main
+
+import (
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/aws"
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"os"
+	"log"
+)
+
+var (
+	connection = ec2.New(session.New())
+)
+
+
+func main() {
+
+	instanceKeyFilter   := os.Getenv("INSTANCE_KEY")
+	instanceValueFilter := os.Getenv("INSTANCE_VALUE")
+
+	// Filter instances with role set to master
+	instanceInput := &ec2.DescribeInstancesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name: 	aws.String(fmt.Sprintf("tag:%s", instanceKeyFilter )),
+				Values: []*string{ aws.String(instanceValueFilter)},
+			},
+		},
+	}
+
+	// A reservation corresponds to a command to start instances
+	// A reservation is what you do to provision instances, while an instance is what you get
+	reservation, err := connection.DescribeInstances(instanceInput)
+	if err != nil {
+		log.Fatal(err.Error())
+	}
+
+	// Iterate over reservations
+	for index := range reservation.Reservations {
+
+		// Iterate reservations to access ec2 instances
+		for _, inst := range reservation.Reservations[index].Instances {
+
+			// Iterate over each instance's block device metadata
+			for _, blk := range inst.BlockDeviceMappings {
+
+				// Exclude root volumes.  Root volumes have "Delete on termination" set to true mostly
+				rootDevice := *blk.Ebs.DeleteOnTermination
+
+				if rootDevice != true {
+					vol_id := blk.Ebs.VolumeId
+
+					snapshotInput := &ec2.CreateSnapshotInput{
+						VolumeId: vol_id,
+					}
+
+					// Perform ebs snapshot operation
+					snapIt, err := connection.CreateSnapshot(snapshotInput)
+					if err != nil {
+						if aerr, ok := err.(awserr.Error); ok {
+							switch aerr.Code() {
+							default:
+								fmt.Println(aerr.Error())
+							}
+						} else {
+							// Print the error, cast err to awserr.Error to get the Code and
+							// Message from an error.
+							fmt.Println(err.Error())
+						}
+						return
```

Comment:
Is `return` what you really want here? (same considerations in the other similar `if` blocks below)

If take the snapshot of one of the instances fail, if we return it will not take a snapshot of the other instances right? I don't know if that's what you want or you may prefer to take the snapshot of the rest of the instances anyway.

Also, if the context here is the execution of a lambda function, do we want this to return silently (that would be seen as "all good/the lambda took the all snapshots of all volumes fine") or do we want this to exit with a non-zero code as well? Unless I'm missing something, these errors will stop the execution silently, but maybe I'm missing something.

Reply:
Not quite, I don't think. All this happens during an iteration, so expected behaviour would be for the `range` to continue looping through the instances if one or more were to fail. I would not want to exit on failure for anything other than `DescribeInstances`. 

The `return` is due to the weird error casting.